src/pyodide/python-entrypoint-helper.js

```diff
@@ -120,51 +115,61 @@ function transformMetadata(metadata) {
-    // each one, save it in Pyodide's file system.
-    let hasRequirements = false;
-    const pythonRequirements = [];
-    const micropipRequirements = [];
-    for (const { name, value } of metadata.globals) {
-      if (value.pythonModule !== undefined) {
-        pyodide.FS.writeFile(`/session/${name}.py`, value.pythonModule, {
-          canOwn: true,
-        });
-      }
-
-      if (value.pythonRequirement !== undefined) {
-        hasRequirements = true;
-        if (!EMBEDDED_PYTHON_PACKAGES.includes(name)) {
-          pythonRequirements.push(name);
-        }
-      }
+export async function setupPackages(pyodide, origMetadata) {
+  // The metadata is a JSON-serialised WorkerBundle (defined in pipeline.capnp).
+  const metadata = transformMetadata(origMetadata);
+
+  initializePackageIndex(pyodide, lockFile);
+
+  // Loop through globals that define Python modules in the metadata passed to our Worker. For
+  // each one, save it in Pyodide's file system.
+  const requirements = [];
+  const pythonRequirements = [];
+  const micropipRequirements = [];
+  for (const { name, value } of metadata.globals) {
+    if (value.pythonModule !== undefined) {
+      pyodide.FS.writeFile(`/session/${name}.py`, value.pythonModule, {
+        canOwn: true,
+      });
     }
 
-    if (hasRequirements) {
-      const micropip = pyodide.pyimport("micropip");
-      if (micropipRequirements.length > 0) {
-        // Micropip and ssl packages are contained in the tarball which is extracted above. This means
-        // we should be able to load micropip directly now.
-        await micropip.install(micropipRequirements);
-      }
-
-      // Apply patches that enable some packages to work.
-      const patches = getPatches();
-      if (micropip.list().has("aiohttp") !== undefined) {
-        pyodide.runPython(patches["aiohttp_fetch_patch.py"]);
+    if (value.pythonRequirement !== undefined) {
+      requirements.push(name);
+      if (!EMBEDDED_PYTHON_PACKAGES.includes(name)) {
+        pythonRequirements.push(name);
       }
     }
+  }
 
+  if (pythonRequirements.length > 0) {
     await pyodide.loadPackage(pythonRequirements);
+  }
+
+  if (micropipRequirements.length > 0) {
+    // Micropip and ssl packages are pre-loaded via the packages tarball. This means
+    // we should be able to load micropip directly now.
+    const micropip = pyodide.pyimport("micropip");
+    await micropip.install(micropipRequirements);
+  }
 
-    return await pyodide.pyimport(metadata.mainModule).fetch(request);
-  },
-};
+  // Apply patches that enable some packages to work.
+  if (requirements.includes("aiohttp")) {
+    const mod = await import("pyodide-internal:patches/aiohttp_fetch_patch.py");
+    pyodide.FS.writeFile(
+      "/lib/python3.11/site-packages/aiohttp_fetch_patch.py",
+      new Uint8Array(mod.default),
+      { canOwn: true }
+    );
+    pyodide.pyimport("aiohttp_fetch_patch");
+  }
+  if (requirements.includes("fastapi")) {
```

Comment:
@dom96 these changes may have messed up your langchain demo depending on whether they have `aiohttp` as a `pythonRequirement`.

I think `micropip.freeze().has("aiohttp")` will always return `true` because we are unconditionally exposing it to the filesystem. We could for now just unconditionally load the aiohttp patch but that will be a performance hit when it isn't needed.

Before we have any sort of beta we'll definitely have to ensure that we don't expose any extra packages that user didn't explicitly request. Another thing to discuss is whether the `pythonRequirement` list needs to include all transitive dependencies or just leafs.

Reply:
`Aiohttp` is not one of the `pythonRequirement`s, it is just `langchain` and `openai`, so it definitely expects the transitive dependencies to be resolved and installed implicitly.

I think that for the demo listing all the transitive dependencies would be too painful, but for the long-term it may make sense to do so. How about for now you check for `langchain` or `openai` in the dependencies and if they are there include the aiohttp patch? You can add a TODO in there so we remember to remove this later.